pkg/cli/cli.go

```diff
@@ -0,0 +1,271 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cli
+
+import (
+	"fmt"
+	"log"
+	"os"
+
+	"sigs.k8s.io/kubebuilder/pkg/plugin"
+	"sigs.k8s.io/kubebuilder/pkg/scaffold"
+	"sigs.k8s.io/kubebuilder/pkg/scaffold/project"
+
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+)
+
+const (
+	noticeColor = "\033[1;36m%s\033[0m"
+)
+
+// Option is a function that can configure the CLI
+type Option func(*CLI) error
+
+// CLI defines the command line structure and interfaces that are used to
+// scaffold kubebuilder project files.
+type CLI struct {
+	commandName    string
+	defProjVersion string
+	cmd            *cobra.Command
+	extraCommands  []*cobra.Command
+	plugins        map[string]plugin.BasePlugin
+}
+
+// New creates a new CLI instance.
+func New(opts ...Option) (*CLI, error) {
+	c := &CLI{
+		commandName:    "kubebuilder",
+		defProjVersion: project.Version2,
+		plugins:        make(map[string]plugin.BasePlugin),
+	}
+	for _, opt := range opts {
+		if err := opt(c); err != nil {
+			return nil, err
+		}
+	}
+
+	if err := c.initialize(); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+// Run runs the CLI.
+func (c CLI) Run() error {
+	return c.cmd.Execute()
+}
+
+// WithCommandName is an Option that sets the CLI's root command name.
+func WithCommandName(name string) Option {
+	return func(c *CLI) error {
+		c.commandName = name
+		return nil
+	}
+}
+
+// WithDefaultProjectVersion is an Option that sets the CLI's default project
+// version. Setting an unknown version will result in an error.
+func WithDefaultProjectVersion(version string) Option {
+	return func(c *CLI) error {
+		c.defProjVersion = version
+		return nil
+	}
+}
+
+// WithPlugins is an Option that sets the CLI's plugins. Duplicate plugin
+// versions will result in an error.
+func WithPlugins(plugins ...plugin.BasePlugin) Option {
+	return func(c *CLI) error {
+		for _, p := range plugins {
+			if _, ok := c.plugins[p.Version()]; ok {
```

Comment:
Since `c.plugins` is accessed by project version below, the key type needs to be `p.Version()`. However `c.plugins` should contain a slice of `Plugin` per key as to not overwrite preceding plugins.